core/vault/vault.go

```diff
@@ -0,0 +1,254 @@
+package vault
+
+import (
+	"bytes"
+	"crypto/aes"
+	"crypto/cipher"
+	"crypto/rand"
+	"crypto/sha512"
+	"encoding/base64"
+	"encoding/json"
+	"errors"
+	"io"
+	"io/ioutil"
+	"net/http"
+
+	"github.com/FleekHQ/space-daemon/core/space/domain"
+	"golang.org/x/crypto/pbkdf2"
+)
+
+type vault struct {
+	vaultAPIURL     string
+	vaultSaltSecret string
+}
+
+type VaultItemType string
+
+// Vault item types
+const (
+	PrivateKeyWithMnemonic VaultItemType = "PrivateKeyWithMnemonic"
+)
+
+type VkVersion string
+
+const (
+	VkVersion1 VkVersion = "V1"
+)
+
+// AES requires key length equal to 16, 24 or 32 bytes
+const vaultKeyLength = 32
+
+type VaultItem struct {
+	ItemType VaultItemType
+	Value    string
+}
+
+type storeVaultRequest struct {
+	Vault string `json:"vault"`
+	Vsk   string `json:"vsk"`
+}
+
+type retrieveVaultRequest struct {
+	Vsk string `json:"vsk"`
+}
+
+type retrieveVaultResponse struct {
+	EncryptedVault string `json:"encryptedVault"`
+}
+
+type Vault interface {
+	Store(uuid string, passphrase string, items []VaultItem) (*storeVaultRequest, error)
+	Retrieve(uuid string, passphrase string) ([]VaultItem, error)
+}
+
+func New(vaultAPIURL string, vaultSaltSecret string) *vault {
+	return &vault{
+		vaultAPIURL:     vaultAPIURL,
+		vaultSaltSecret: vaultSaltSecret,
+	}
+}
+
+func (v *vault) Store(uuid string, passphrase string, items []VaultItem) (*storeVaultRequest, error) {
+	// Generate vault file
+	vf, err := json.Marshal(items)
+	if err != nil {
+		return nil, err
+	}
+
+	// Compute vault key
+	vk := v.computeVk(uuid, passphrase, VkVersion1)
+
+	// Encrypt vault file using vault key
+	encVf, err := encrypt(vf, vk)
+	if err != nil {
+		return nil, err
+	}
+
+	// Compute vault service key
+	vsk := v.computeVsk(vk, passphrase, VkVersion1)
+
+	// Submit encrypted file and vsk to vault service
+	storeRequest := &storeVaultRequest{
+		Vault: base64.RawStdEncoding.EncodeToString(encVf),
+		Vsk:   base64.RawStdEncoding.EncodeToString(vsk),
+	}
+	reqJSON, err := json.Marshal(storeRequest)
+	if err != nil {
+		return nil, err
+	}
+
+	resp, err := http.Post(
+		v.vaultAPIURL+"/vaults",
+		"application/json",
+		bytes.NewBuffer(reqJSON),
+	)
+	if err != nil {
+		return nil, err
+	}
+	defer resp.Body.Close()
+
+	_, err = parseAPIResponse(resp)
+	if err != nil {
+		return nil, err
+	}
+
+	return storeRequest, nil
+}
+
+func (v *vault) Retrieve(uuid string, passphrase string) ([]VaultItem, error) {
+	// Compute vault key
+	vk := v.computeVk(uuid, passphrase, VkVersion1)
+
+	// Compute vault service key
+	vsk := v.computeVsk(vk, passphrase, VkVersion1)
+
+	// Send retrieve request to vault service
+	reqJSON, err := json.Marshal(&retrieveVaultRequest{
+		Vsk: base64.RawStdEncoding.EncodeToString(vsk),
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	resp, err := http.Post(
+		v.vaultAPIURL+"/vaults/"+uuid,
```

Comment:
I assume uuid is obtained by `getIdentityByUsername`, right? Maybe it'd be better if we can use username here directly, so we can secure `getIdentityByUsername` endpoint with app JWT (rate limits on token) and this one differently (like IP-rate limited).

Reply:
Don't want to use username for the following reasons:
- Vault endpoint shouldn't change if the user changes their username
- Daemon doesn't deal with usernames. It shouldn't deal with uuids, but at least these are immutable registries that map to a vault. So if we migrate the identity provider to something like Ceramic later on, we can keep using these uuids as an attached document.

I don't see any strong reasons for securing the `getIdentityByUsername` endpoint. These are public registries anyways 🤔

Reply:
Oh I see, ok. Initially I thought securing that endpoint would make harder to scan users, like you can have a list of twitter accounts (github, etc.) and you can scan our endpoint and reveal that user `amanda93` is using space storage. Then you can use this information for phishing attacks and so on (aka "Your documents leaked from your space storage!")..

Reply:
Gotcha, I see what you are saying. On the other hand, it's very easy for this scammer to create an account first and then use their JWT before calling the address book, so this would be a very small barrier.